src/components/image/index.tsx

```diff
@@ -46,44 +47,70 @@ export function Image({
   maxSize,
   alt,
   inputProps,
+  errorComponent: ErrorComponent,
   ...imgOptions
 }: Props): ReactElement {
+  const [error, setError] = useState<null | Error>(null);
   const { fetch } = useContext(SessionContext);
 
   const thingContext = useContext(ThingContext);
   const { thing: contextThing } = thingContext;
 
   const thing = propThing || contextThing;
 
-  if (!thing) {
-    throw new Error("Thing not found as property or in context");
+  if (!thing && !error) {
+    const thingNotFound = new Error(
+      "Thing not found as property or in context"
+    );
+
+    if (!ErrorComponent && !onError) {
+      throw thingNotFound;
+    } else if (onError) {
+      onError(thingNotFound);
+    }
+
+    setError(thingNotFound);
   }
 
-  const src = getUrl(thing, property);
+  const src = thing ? getUrl(thing, property) : null;
+
+  if (!src && !error) {
+    const urlNotFound = new Error("URL not found for given property");
+
+    if (!ErrorComponent && !onError) {
+      throw urlNotFound;
+    } else if (onError) {
+      onError(urlNotFound);
+    }
 
-  if (!src) {
-    throw new Error("URL not found for given property");
+    setError(urlNotFound);
   }
 
   const [imgObjectUrl, setImgObjectUrl] = useState("");
 
   useEffect(() => {
-    retrieveFile(src, fetch)
-      .then(setImgObjectUrl)
-      .catch((error) => {
-        if (onError) {
-          onError(error);
-        } else {
-          setImgObjectUrl(() => {
-            throw error;
-          });
-        }
-      });
-  }, [src, onError, fetch]);
+    if (src) {
+      retrieveFile(src, fetch)
```

Comment:
Maybe it's time to extract the data fetching into a hook? Or alternatively, to split this into a component that does the fetching and one that does the rendering?

Reply:
A `const { data, error, inProgress } = useFile(property, thing?, dataset)` that pulled thing/dataset from context would be great. I'll make a ticket.